src/Datadog.Trace/Propagation/W3CSpanContextPropagator.cs

```diff
@@ -0,0 +1,270 @@
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Text;
+using Datadog.Trace.Conventions;
+using Datadog.Trace.Logging;
+
+namespace Datadog.Trace.Propagation
+{
+    internal class W3CSpanContextPropagator : IPropagator
+    {
+        private const string TraceParentFormat = "00-{0}-{1}-01";
+
+        // The following length limits are from Trace Context v1 https://www.w3.org/TR/trace-context-1/#key
+        private const int TraceStateKeyMaxLength = 256;
+        private const int TraceStateKeyTenantMaxLength = 241;
+        private const int TraceStateKeyVendorMaxLength = 14;
+        private const int TraceStateValueMaxLength = 256;
+
+        private static readonly IDatadogLogger Log = DatadogLogging.GetLoggerFor<W3CSpanContextPropagator>();
+        private static readonly int VersionPrefixIdLength = "00-".Length;
+        private static readonly int TraceIdLength = "0af7651916cd43dd8448eb211c80319c".Length;
+        private static readonly int VersionAndTraceIdLength = "00-0af7651916cd43dd8448eb211c80319c-".Length;
+        private static readonly int SpanIdLength = "00f067aa0ba902b7".Length;
+
+        private readonly ITraceIdConvention _traceIdConvention;
+
+        public W3CSpanContextPropagator(ITraceIdConvention traceIdConvention)
+        {
+            _traceIdConvention = traceIdConvention;
+        }
+
+        public void Inject<T>(SpanContext context, T carrier, Action<T, string, string> setter)
+        {
+            // lock sampling priority when span propagates.
+            context.TraceContext?.LockSamplingPriority();
+
+            setter(carrier, W3CHeaderNames.TraceParent, string.Format(TraceParentFormat, context.TraceId.ToString(), context.SpanId.ToString("x16")));
+            if (!string.IsNullOrEmpty(context.TraceState))
+            {
+                setter(carrier, W3CHeaderNames.TraceState, context.TraceState);
+            }
+        }
+
+        public SpanContext Extract<T>(T carrier, Func<T, string, IEnumerable<string>> getter)
+        {
+            var traceStateCollection = getter(carrier, W3CHeaderNames.TraceState);
+            var traceState = ExtractTraceState(traceStateCollection);
+
+            var traceParentValues = getter(carrier, W3CHeaderNames.TraceParent).ToList();
```

Comment:
In the SDK, they get the `Count()` of the resulting enumerable and then access the traceparent by getting the `First()` of the resulting enumerable. That seems slightly more efficient so we don't allocate a new List when we don't really need to

Reply:
Hmm, this could cause multiple enumeration, so I would rather not do that.

Reply:
I think Zach's point is valid. `string.Join(",", traceParentValues) ` could be just `traceParentHeader`, cause `traceParentValues.Count != 1` will not let you past anyway. So you are left with `Count()` and `First()` operations - which means full list will be never needed.

Reply:
... or even better, `FirstOrDefault()` single enumerable operation.

Reply:
The FirstOrDefault would be cool but then we have no way of checking if there was more than one value, and for now in this case we return null SpanContext.

Reply:
sorry, I meant `SingleOrDefault()`, my bad.